modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -137,6 +150,153 @@ public IgniteRel visit(IgniteTableScan rel) {
+            @Nullable List<RexNode> projections,
+            @Nullable List<List<RexNode>> exprProjections,
+            RexBuilder rexBuilder
+    ) {
+        if (vals == null && exprProjections == null) {
+            return;
+        }
+
+        IgniteDistribution distribution = table.distribution();
+        if (!distribution.function().affinity()) {
+            return;
+        }
+
+        IntArrayList keysList = new IntArrayList(distribution.getKeys().size());
+        for (Integer key : distribution.getKeys()) {
+            keysList.add(key.intValue());
+        }
+
+        Mappings.TargetMapping mapping = null;
+
+        if (projections != null) {
+            IntArrayList projList = new IntArrayList();
+            for (RexNode node : projections) {
+                if (node instanceof RexInputRef) {
+                    int idx = ((RexInputRef) node).getIndex();
+                    if (keysList.contains(idx)) {
+                        projList.add(((RexInputRef) node).getIndex());
+                    }
+                } else {
+                    if (!isValueExpr(node)) {
+                        return;
+                    }
+                }
+            }
+
+            if (!projList.containsAll(keysList)) {
+                return;
+            }
+
+            projections = replaceInputRefs(projections);
+
+            mapping = RexUtils.inversePermutation(projections,
+                    table.getRowType(Commons.typeFactory()), true);
+        }
+
+        List<RexNode> andEqNodes = new ArrayList<>();
+
+        RelDataType rowTypes = table.getRowType(Commons.typeFactory());
+
+        List<List<RexNode>> dataRows = vals != null ? Commons.cast(vals.getTuples()) : exprProjections;
+
+        for (List<RexNode> items : dataRows) {
+            List<RexNode> andNodes = new ArrayList<>(keysList.size());
+            for (int key : keysList) {
+                RexLocalRef ref;
+                if (projections != null) {
+                    ref = (RexLocalRef) projections.get(mapping.getSourceOpt(key));
+                } else {
+                    ref = rexBuilder.makeLocalRef(rowTypes.getFieldList().get(key).getType(), key);
+                }
+                RexNode lit = items.get(mapping != null ? mapping.getSourceOpt(key) : key);
+                if (!isValueExpr(lit)) {
+                    return;
+                }
+                RexNode eq = rexBuilder.makeCall(SqlStdOperatorTable.EQUALS, ref, lit);
+                andNodes.add(eq);
+            }
+            if (andNodes.size() > 1) {
+                assert andNodes.size() % 2 == 0;
+                RexNode node0 = rexBuilder.makeCall(SqlStdOperatorTable.AND, andNodes);
+                andEqNodes.add(node0);
+            } else {
+                andEqNodes.add(andNodes.get(0));
+            }
+        }
+
+        if (!nullOrEmpty(andEqNodes)) {
+            RexNode call = rexBuilder.makeCall(SqlStdOperatorTable.OR, andEqNodes);
```

Comment:
in case of single-tuple insert exception is thrown here

```
java.lang.AssertionError: wrong operand count 1 for OR

	at org.apache.calcite.util.Litmus.lambda$static$0(Litmus.java:31)
	at org.apache.calcite.sql.SqlBinaryOperator.validRexOperands(SqlBinaryOperator.java:227)
	at org.apache.calcite.rex.RexCall.<init>(RexCall.java:85)
	at org.apache.calcite.rex.RexBuilder.makeCall(RexBuilder.java:272)
	at org.apache.ignite.internal.sql.engine.prepare.pruning.PartitionPruningMetadataExtractor.extractFromValues(PartitionPruningMetadataExtractor.java:290)
```

we should add such tests

Reply:
add test, fixed